### PR TITLE
Automatic OSPF router-id assignment.

### PR DESCRIPTION
**Pull request by xschily**
Tags: fix, dev, Github issue #19

Change-Id: I70fff570e788dbc7025870c757a3a4baaa2e9f0d
Signed-off-by: Ilya Schepin ischepin@mera.ru

There are certain rules to assign OSPF (and BGP) router-ids automatically in case the user did not select the id manually. When id is assigned automatically, OSPF will pick it from zebra using zclient API (1st issue, according to OpenSwitch architecture there must be no interaction between components except interaction through OVSDB). OSPF will not write router-id retrieved from zebra to OSPF_Router table (2nd issue, CLI will print everyone except OSPF_Router:router_id as OSPF neighbor in `show ip ospf neighbor` command).  

Router-id should be obtained from OVSDB as BGP does it; using OVSDB table VRF, column active_router_id. Router-id should be written to OVSDB table OSPF_Router by OSPFD when assigned automatically.

### CODE REVIEW

File: ospfd/ospf_ovsdb_if.c [hunk 1]
```diff
@@ -968,31 +969,76 @@ int ovsdb_options_count(u_char options_val)
 }
 
 int
+ospf_set_router_id_from_vrf (const struct ovsrec_ospf_router *ovs_ospf_router,
```
Comment (stroykov) [inline ospfd/ospf_ovsdb_if.c]: rename to "ospf_router_ovsdb_router_id_update"

File: ops-tests/component/ospf/test_ospfv2_ct_verify_config.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    step('### Test to verify correct OSPF router-id assignment ###')
+    step('### 1. Test that router-id is assigned automatically from VRF table ###')
+    sw("configure terminal")
+    sw("interface loopback 1")
+    sw("ip address %s/24" % ospf_auto_router_id)
+    sw("exit")
+    sw("configure terminal")
+    sw("interface loopback 2")
+    sw("ip address %s/24" % lo2_addr)
```
Comment (stroykov) [inline ops-tests/component/ospf/test_ospfv2_ct_verify_config.py]: /32

File: ops-tests/component/ospf/test_ospfv2_ct_verify_config.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    ospf_auto_router_id = "192.168.1.1"
+    ospf_man_router_id = "1.1.1.1"
+    lo2_addr = "172.16.1.1"
+    assert sw is not None
+
+    step('### Test to verify correct OSPF router-id assignment ###')
+    step('### 1. Test that router-id is assigned automatically from VRF table ###')
+    sw("configure terminal")
+    sw("interface loopback 1")
+    sw("ip address %s/24" % ospf_auto_router_id)
```
Comment (stroykov) [inline ops-tests/component/ospf/test_ospfv2_ct_verify_config.py]: /32

**Review by stroykov — CHANGES_REQUESTED** [review]

**Comment by vkarpov** [comment]
No more comments.

**Review by apolovin — APPROVED** [review]
Ok for me

**Pull request closed without merging**